src/script/miniscript.h

```diff
@@ -0,0 +1,998 @@
+                } else if (in[j] == 'l') {
+                    // The l: wrapper is equivalent to or_i(0,X)
+                    constructed.push_back(MakeNodeRef<Key>(NodeType::JUST_0));
+                    to_parse.emplace_back(ParseContext::OR_I, -1, -1);
+                } else {
+                    return {};
+                }
+            }
+            to_parse.emplace_back(ParseContext::EXPR, -1, -1);
+            in = in.subspan(colon_index + 1);
+            break;
+        }
+        case ParseContext::EXPR: {
+            if (Const("0", in)) {
+                constructed.push_back(MakeNodeRef<Key>(NodeType::JUST_0));
+            } else if (Const("1", in)) {
+                constructed.push_back(MakeNodeRef<Key>(NodeType::JUST_1));
+            } else if (Const("pk(", in)) {
+                Key key;
+                int key_size = FindNextChar(in, ')');
+                if (key_size < 1) return {};
+                if (!ctx.FromString(in.begin(), in.begin() + key_size, key)) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::WRAP_C, Vector(MakeNodeRef<Key>(NodeType::PK_K, Vector(std::move(key))))));
+                in = in.subspan(key_size + 1);
+            } else if (Const("pkh(", in)) {
+                Key key;
+                int key_size = FindNextChar(in, ')');
+                if (key_size < 1) return {};
+                if (!ctx.FromString(in.begin(), in.begin() + key_size, key)) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::WRAP_C, Vector(MakeNodeRef<Key>(NodeType::PK_H, Vector(std::move(key))))));
+                in = in.subspan(key_size + 1);
+            } else if (Const("pk_k(", in)) {
+                Key key;
+                int key_size = FindNextChar(in, ')');
+                if (key_size < 1) return {};
+                if (!ctx.FromString(in.begin(), in.begin() + key_size, key)) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::PK_K, Vector(std::move(key))));
+                in = in.subspan(key_size + 1);
+            } else if (Const("pk_h(", in)) {
+                Key key;
+                int key_size = FindNextChar(in, ')');
+                if (key_size < 1) return {};
+                if (!ctx.FromString(in.begin(), in.begin() + key_size, key)) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::PK_H, Vector(std::move(key))));
+                in = in.subspan(key_size + 1);
+            } else if (Const("sha256(", in)) {
+                int hash_size = FindNextChar(in, ')');
+                if (hash_size < 1) return {};
+                std::string val = std::string(in.begin(), in.begin() + hash_size);
+                if (!IsHex(val)) return {};
+                auto hash = ParseHex(val);
+                if (hash.size() != 32) return {};
+                constructed.push_back(MakeNodeRef<Key>(NodeType::SHA256, std::move(hash)));
+                in = in.subspan(hash_size + 1);
```

Comment:
In 3a2c00d9bc3c7a34fb8727d76e9c3a691395b67c "Miniscript: type system, script creation, text notation, tests"

Can the hex data parsing stuff be refactored into a function instead of copy and pasted for the next 3 expressions?